src/core/query.ts

```diff
@@ -33,18 +33,36 @@ interface QueryConfig<TData, TError, TQueryFnData> {
 
 export interface QueryState<TData = unknown, TError = unknown> {
   data: TData | undefined
+  dataOrigin?: QueryDataOrigin
   dataUpdateCount: number
+  dataUpdatedAt: number
   error: TError | null
+  errorOrigin?: QueryErrorOrigin
   errorUpdateCount: number
-  failureCount: number
+  errorUpdatedAt: number
+  fetchFailureCount: number
   fetchMeta: any
+  fetchOrigin?: QueryFetchOrigin
   isFetching: boolean
   isInvalidated: boolean
   isPaused: boolean
   status: QueryStatus
-  updatedAt: number
 }
 
+export type QueryFetchOrigin =
+  | 'clientFetch'
+  | 'clientRefetch'
+  | 'observerFetch'
+  | 'observerRefetch'
+  | 'observerRefetchInterval'
+  | 'observerRefetchOnMount'
+  | 'observerRefetchOnReconnect'
+  | 'observerRefetchOnWindowFocus'
+
+export type QueryDataOrigin = QueryFetchOrigin | 'initialData' | 'setData'
```

Comment:
quick question: what about `placeholderData`? Wouldn’t that also be a possible origin like `initialData`?

Reply:
Not on the query level at least as the placeholder data will never be persisted in the query. Additional origins could be added on the observer level though...

Reply:
should hydration count as its own origin, or should it just use the origin that was used to load it on the server?

Reply:
Think it should hydrate the original origins, otherwise the UI might change between hydrations.